Various Win32 API additions/fixes

aguinet opened the pull request:
This PR contains various Win32 API additions/fixes:
* add a get_size method to Miasm heap object, which allows the
implementation of mscvrt_realloc
* add the concept of "current directory", with the default value being
arbitrary set to "c:\tmp", which allows the implementation of
{Get,Set}CurrentDirecrtory

Various other methods implemented:

 - advapi32_RegCloseKey
 - advapi32_RegCreateKeyW
 - advapi32_RegSetValueExA
 - advapi32_RegSetValueExW
 - kernel32_GetProcessHeap
 - msvcrt_delete
 - msvcrt_fprintf
 - msvcrt_fwrite
 - msvcrt__mbscpy
 - msvcrt_new
 - msvcrt_realloc
 - msvcrt_sprintf
 - msvcrt_srand
 - msvcrt_strrchr
 - msvcrt_swprintf
 - msvcrt_wcscat
 - msvcrt_wcscmp
 - msvcrt_wcscpy
 - msvcrt__wcsicmp
 - msvcrt_wcslen
 - msvcrt_wcsncpy
 - msvcrt__wcsnicmp
 - msvcrt_wcsrchr

commial commented on miasm2/os_dep/common.py:
```diff
@@ -72,10 +73,14 @@ def vm_alloc(self, vm, size, perm=PAGE_READ | PAGE_WRITE):
-        addr = self.next_addr(size)
-        vm.add_memory_page(addr, perm, "\x00" * size,
+        addr = self.next_addr(size+8)
+        vm.add_memory_page(addr, perm, "\x00" * (size+8),
```
Mixing a meta-info such as the page size with the data is probably not a good idea.
This is more a "structural" information, such as memory access rights, and it is already save in the implementation of the memory pages.
For now, there is no direct API to get it back, but you can obtain it by using `vm.get_all_memory()[start_addr]["size"]`.
An API like the one with mem access is maybe missing.

aguinet replied:
The API you give is linked to memory pages, and not the heap that is used here no? This would mean that the size returned by ``vm.get_all_memory()[start_addr]["size"]`` would be the size of the page where the heap resides, but not the size of the allocated heap chunk.

commial commented on miasm2/os_dep/win_api_x86_32.py:
```diff
@@ -1411,6 +1456,16 @@ def kernel32_lstrcpyA(jitter):
+def msvcrt__mbscpy(jitter):
+    ret_ad, args = jitter.func_args_cdecl(["ptr_str1", "ptr_str2"])
+    s2 = jitter.get_str_unic(args.ptr_str2)
+    print("mbscpy '%s'" % s2)
```
You have made an effort to use logging facilities everywhere else, too bad to use a `print` here

aguinet replied:
It's because I missed this one :)

commial commented on miasm2/os_dep/win_api_x86_32.py:
```diff
@@ -1963,6 +2044,72 @@ def user32_IsCharAlphaNumericA(jitter):
     jitter.func_ret_stdcall(ret_ad, ret)
 
 
+def get_fmt_args(jitter, fmt, cur_arg, get_str, char_size=1):
```
Instead of copying `get_fmt_args` from `linux_stdlib`, it would be probably better to have them in common in `os_dep/common.py`; this way, they will both benefit from future update and fix...

aguinet replied:
Right, this is an extract from a previous "work session" with miasm. I forgot that I copy-pasted it from `linux_stdlib`

aguinet commented on miasm2/os_dep/win_api_x86_32.py:
```diff
@@ -1963,6 +2044,72 @@ def user32_IsCharAlphaNumericA(jitter):
+    ret = len(output)
+    log.info("fprintf(%x, '%s') = '%s'" % (args.file, jitter.get_str_ansi(args.fmt), output))
+
+    fd = upck32(jitter.vm.get_mem(args.file + 0x10, 4))
```
There is `x86_32` in the filename :) ?

commial replied:
Yeah, it is misleading. But for now, API implementation are shared between x86_32 and x86_64, thanks to "abstraction" like calling conventions wrapper

aguinet replied:
It's actually done in many other places like that :)

commial replied:
Ok for me

aguinet commented:
@commial I fixed almost everything except the 32-bit FD "issue" here https://github.com/cea-sec/miasm/pull/616/commits/789a4a937280b07fdbd6855cc7208c2b24c4f56c , and added some related tests :)

aguinet commented:
If that's fine for you I'll squash this!

commial commented on miasm2/os_dep/common.py:
```diff
@@ -73,14 +73,13 @@ def vm_alloc(self, vm, size, perm=PAGE_READ | PAGE_WRITE):
-        addr = self.next_addr(size+8)
-        vm.add_memory_page(addr, perm, "\x00" * (size+8),
+        addr = self.next_addr(size)
+        vm.add_memory_page(addr, perm, "\x00" * (size),
```
Tiny remark: you left your parenthesis around `(size)`, which lead to a diff on this function for the whole PR, where there is no need to

aguinet commented:
@commial I removed the "bad" diff, squashed everything into one commit and rebased on master. I also added a test for ``msvcrt_realloc`` (so that ``heap.get_size`` is tested as a side effect), and also removed ``kernel32_GetProcessHeap`` which was a "return 0xdeadbeef"-only implementation..

commial commented on miasm2/os_dep/common.py:
```diff
@@ -73,14 +73,13 @@ def vm_alloc(self, vm, size, perm=PAGE_READ | PAGE_WRITE):
 
     def get_size(self, vm, ptr):
-        return upck64(vm.get_mem(ptr-8, 8))
+        return vm.get_all_memory()[ptr]["size"]
```
As you add a new public function, can you add a docstring describing it?
I know this is not the case for every function in this module, but we (at least try to) add docstring on new implemented function. Thus, Miasm readibility gets better on average :smiley: 

For this function, there is a particularity: `ptr` must be the base address of the mapped function. If you want to have a more permissive function, you can do something like (untested but should work):
```Python
assert vm.is_mapped(ptr, 1)
data = vm.get_all_memory()
ptr_page = data.get(ptr, None)
if ptr_page is None:
    for address, page_info in data.iteritems():
        if address <= ptr < address + page_info["size"]:
            ptr_page = page_info
            break
    else:
        raise RuntimeError("Must never happen (unmapped but mark as mapped by API)")
return ptr_page["size"]
```
Note for a future PR: this function should probably reside in VmMngr implementation

commial commented on miasm2/os_dep/win_api_x86_32.py:
```diff
@@ -1793,13 +1848,39 @@ def msvcrt_memset(jitter):
 def msvcrt_memcpy(jitter):
     ret_ad, args = jitter.func_args_cdecl(['dst', 'src', 'size'])
     s = jitter.vm.get_mem(args.src, args.size)
+    #log.info("memcpy buf %s" % s.encode("hex"))
```
You still have a commented log here, you can uncomment it

aguinet replied:
Or just remove the line :) ?

commial replied:
Your call

commial commented on miasm2/os_dep/common.py:
```diff
@@ -73,10 +74,13 @@ def vm_alloc(self, vm, size, perm=PAGE_READ | PAGE_WRITE):
         return addr
 
+    def get_size(self, vm, ptr):
+        return vm.get_all_memory()[ptr]["size"]
```
As you add a new public function, can you add a docstring describing it?
I know this is not the case for every function in this module, but we (at least try to) add docstring on new implemented function. Thus, Miasm readibility gets better on average :smiley: 

For this function, there is a particularity: `ptr` must be the base address of the mapped function. If you want to have a more permissive function, you can do something like (untested but should work):
```Python
assert vm.is_mapped(ptr, 1)
data = vm.get_all_memory()
ptr_page = data.get(ptr, None)
if ptr_page is None:
    for address, page_info in data.iteritems():
        if address <= ptr < address + page_info["size"]:
            ptr_page = page_info
            break
    else:
        raise RuntimeError("Must never happen (unmapped but mark as mapped by API)")
return ptr_page["size"]
```
Note for a future PR: this function should probably reside in VmMngr implementation

aguinet replied:
I think the semantic of the function should be indeed more clear. About what you propose, I think we should actually make 2 functions : ``get_page_size``, which asserts that the ``ptr`` is the start address of a page, and a ``get_alloc_size``, which does what you do and returns the number of available bytes starting from ``ptr`` (so that's actually a bit different from your implementation, but as long as it's properly documented).  What do you think?

commial replied:
It could be confusing (between `get_page_size` / `get_alloc_size`). In addition, we do not have any use of a `get_alloc_size` as you describe for now.
So I'll prefer a unique `get_page_size` API. If you find it confusing to be able to get the allocated page size for a address in the middle of the page, drop my comment but make it clear in the docstring (something like "@ptr: starting address of the target page")

aguinet replied:
Hmm indeed the usage of ``get_alloc_size`` is really limited. Your proposal is fine to me as long as it's also stated in the documentation. I'll do it this way thus :)

commial commented:
Good job! I only have a few remarks left and it will be OK for me.

One last comment: adding test is a very good thing for the project's health, but when you add new test file, they must be tracked (to be launched) by `test/test_all.py`.
So you can add [here](https://github.com/cea-sec/miasm/blob/master/test/test_all.py#L269): `os.path.join("linux", "stdlib.py"),` and travis will launch it :+1:

aguinet commented:
Okay I'll do that :)

aguinet commented:
@commial I made a commit with the latest remarks. Let me know if that's okay! (and I'll squash again :))

commial commented:
Thanks again for this contrib!

(I've squashed the 2 commits ;))

aguinet commented:
Thanks :) !
